src/tree/gpu_hist/gradient_based_sampler.cu

```diff
@@ -0,0 +1,225 @@
+/*!
+ * Copyright 2019 by XGBoost Contributors
+ */
+#include <thrust/functional.h>
+#include <thrust/random.h>
+#include <thrust/transform.h>
+#include <xgboost/host_device_vector.h>
+#include <xgboost/logging.h>
+
+#include <algorithm>
+
+#include "../../common/compressed_iterator.h"
+#include "../../common/random.h"
+#include "gradient_based_sampler.cuh"
+
+namespace xgboost {
+namespace tree {
+
+GradientBasedSampler::GradientBasedSampler(BatchParam batch_param,
+                                           EllpackInfo info,
+                                           size_t n_rows,
+                                           float subsample,
+                                           SamplingMethod sampling_method)
+    : batch_param_(batch_param), info_(info), sampling_method_(sampling_method) {
+  monitor_.Init("gradient_based_sampler");
+
+  // If `subsample` is not specified, try to figure out how many rows to sample based on available
+  // free GPU memory.
+  if (subsample == 0.0f || subsample == 1.0f) {
+    sample_rows_ = MaxSampleRows(n_rows);
+  } else {
+    sample_rows_ = n_rows * subsample;
```

Comment:
what if there isn't enough memory to honor the user-defined subsample % of rows? should this instead be the following in *addition* to this?

`sample_rows_ = std::min(sample_rows_, MaxSampleRows(sample_rows_));`

followed by some info message saying that the user thresholds aren't being honored, if it is being scaled down?

Reply:
As @RAMitchell suggested, maybe we should just get rid of guessing available memory and let user set `subsample` explicitely.